File: src/main/java/cn/dsnbo/bedrockplayersupport/BedrockPlayerSupport.java
package cn.dsnbo.bedrockplayersupport;

import cn.dsnbo.bedrockplayersupport.command.HomeFormCommand;
import cn.dsnbo.bedrockplayersupport.command.MsgFormCommand;
import cn.dsnbo.bedrockplayersupport.command.TpFormCommand;
import cn.dsnbo.bedrockplayersupport.config.Config;
import cn.dsnbo.bedrockplayersupport.config.ConfigManager;
import cn.dsnbo.bedrockplayersupport.config.Language;
import cn.dsnbo.bedrockplayersupport.form.MainForm;
import cn.dsnbo.bedrockplayersupport.listener.PlayerListener;
import cn.dsnbo.bedrockplayersupport.listener.login.CatSeedLoginListener;
import cn.dsnbo.bedrockplayersupport.listener.login.NexAuthListener;
import cn.dsnbo.bedrockplayersupport.listener.login.OtherAuthListener;
import cn.dsnbo.bedrockplayersupport.listener.teleport.HuskHomesTeleportListener;
import cn.dsnbo.bedrockplayersupport.listener.login.AuthMeListener;
import cn.dsnbo.bedrockplayersupport.listener.teleport.CMITeleportListener;
import cn.dsnbo.bedrockplayersupport.listener.teleport.EssXTeleportListener;
import cn.dsnbo.bedrockplayersupport.util.Update;
import java.io.File;
import java.util.Locale;
import com.github.Anon8281.universalScheduler.UniversalScheduler;
import com.github.Anon8281.universalScheduler.scheduling.schedulers.TaskScheduler;
import lombok.Getter;
import org.bukkit.Bukkit;
import org.bukkit.plugin.java.JavaPlugin;
import org.geysermc.floodgate.api.FloodgateApi;

/**
 * @author DongShaoNB
 */
public final class BedrockPlayerSupport extends JavaPlugin {

  @Getter
  private static BedrockPlayerSupport instance;
  @Getter
  private static FloodgateApi floodgateApi;
  @Getter
  private static BasicPlugin basicPlugin;
  @Getter
  private static ConfigManager<Config> mainConfigManager;
  @Getter
  private static ConfigManager<Language> languageConfigManager;
  @Getter
  private static String language;
  @Getter
  private static String systemLanguage;
  @Getter
  private static TaskScheduler scheduler;
  private boolean isCmiEnabled;
  private boolean isEssxEnabled;
  private boolean isHuskHomesEnabled;
  private boolean isAuthMeEnabled;
  private boolean isCatSeedLoginEnabled;
  private boolean isNexAuthEnabled;
  private boolean isFloodgateEnabled;



  @Override
  public void onEnable() {
    instance = this;
    scheduler = UniversalScheduler.getScheduler(this);
    floodgateApi = FloodgateApi.getInstance();
    systemLanguage = Locale.getDefault().getLanguage();
    loadConfig();
    loadDependLoadStatus();
    printSupportPluginLoadStatus();
    registerBasicListener();
    registerAuthListener();
    loadMetrics();
    new MainForm();
  }

  @Override
  public void onDisable() {

  }

  private void loadDependLoadStatus() {
    isCmiEnabled = !(Bukkit.getPluginManager().getPlugin("CMI") == null);
    isEssxEnabled = !(Bukkit.getPluginManager().getPlugin("Essentials") == null);
    isHuskHomesEnabled = !(Bukkit.getPluginManager().getPlugin("HuskHomes") == null);
    isAuthMeEnabled = !(Bukkit.getPluginManager().getPlugin("AuthMe") == null);
    isCatSeedLoginEnabled = !(Bukkit.getPluginManager().getPlugin("CatSeedLogin") == null);
    isNexAuthEnabled = !(Bukkit.getPluginManager().getPlugin("NexAuth") == null);
    isFloodgateEnabled = !(Bukkit.getPluginManager().getPlugin("floodgate") == null);
  }

  public void loadConfig() {
    File langDirectory = new File(getDataFolder(), "/lang/");
    if (!langDirectory.exists()) {
      langDirectory.mkdir();
    }
    mainConfigManager = ConfigManager.create(getDataFolder().toPath(), "config.yml", Config.class);
    mainConfigManager.reloadConfig();
    Config config = getMainConfigManager().getConfigData();
    language = config.language();
    saveResource("/lang/" + language + ".yml", false);
    languageConfigManager = ConfigManager.create(langDirectory.toPath(), language + ".yml",
        Language.class);
    languageConfigManager.reloadConfig();
    if (config.enableTeleportForm()) {
      getCommand("tpgui").setExecutor(new TpFormCommand());
    }
    if (config.enableMsgForm()) {
      getCommand("msggui").setExecutor(new MsgFormCommand());
    }
    if (config.enableHomeForm()) {
      getCommand("homegui").setExecutor(new HomeFormCommand());
    }
    if (config.enableBackForm()) {
      Bukkit.getPluginManager().registerEvents(new PlayerListener(), this);
    }
    if (config.checkUpdate()) {
      Update.checkUpdate(latestVersion -> {
        String currentVersion = getDescription().getVersion();
        if (currentVersion.equals(latestVersion)) {
          if ("zh".equalsIgnoreCase(systemLanguage)) {
            getLogger().info("插件是最新版本，继续保持~");
          } else {
            getLogger().info("The plugin is the latest version");
          }
        } else {
          if ("zh".equalsIgnoreCase(systemLanguage)) {
            getLogger().info("有新版本可以更新!");
            getLogger().info("当前版本: " + currentVersion + " | 最新版本: " + latestVersion);
          } else {
            getLogger().info("There is a new version that can be updated!");
          getLogger().info("Current version: " + currentVersion + " | Latest version: " + latestVersion);
          }
        }
      });
    }
  }

  private void registerBasicListener() {
    switch (getMainConfigManager().getConfigData().basicPlugin().toLowerCase()) {
      case "cmi" -> {
        if (isCmiEnabled) {
          basicPlugin = BasicPlugin.CMI;
          Bukkit.getPluginManager().registerEvents(new CMITeleportListener(), this);
        } else {
          basicPlugin = BasicPlugin.NONE;
        }
      }
      case "essentialsx" -> {
        if (isEssxEnabled) {
          basicPlugin = BasicPlugin.EssentialsX;
          Bukkit.getPluginManager().registerEvents(new EssXTeleportListener(), this);
        } else {
          basicPlugin = BasicPlugin.NONE;
        }
      }
      case "huskhomes" -> {
        if (isHuskHomesEnabled) {
          basicPlugin = BasicPlugin.HuskHomes;
          Bukkit.getPluginManager().registerEvents(new HuskHomesTeleportListener(), this);
        } else {
          basicPlugin = BasicPlugin.NONE;
        }
      }
      case "none" -> {
        basicPlugin = BasicPlugin.NONE;
      }
      default -> {
        if (isCmiEnabled) {
          Bukkit.getPluginManager().registerEvents(new CMITeleportListener(), this);
          basicPlugin = BasicPlugin.CMI;
        } else if (isEssxEnabled) {
          Bukkit.getPluginManager().registerEvents(new EssXTeleportListener(), this);
          basicPlugin = BasicPlugin.EssentialsX;
        } else if (isHuskHomesEnabled) {
          Bukkit.getPluginManager().registerEvents(new HuskHomesTeleportListener(), this);
          basicPlugin = BasicPlugin.HuskHomes;
        } else {
          basicPlugin = BasicPlugin.NONE;
        }
      }
    }
  }


  private void registerAuthListener() {
    if (getMainConfigManager().getConfigData().enableLogin()
        || getMainConfigManager().getConfigData().enableRegister()) {
      switch (getMainConfigManager().getConfigData().authPlugin().toLowerCase()) {
        case "authme" -> {
          if (isAuthMeEnabled) {
            Bukkit.getPluginManager().registerEvents(new AuthMeListener(), this);
          }
        }
        case "catseedlogin" -> {
          if (isCatSeedLoginEnabled) {
            Bukkit.getPluginManager().registerEvents(new CatSeedLoginListener(), this);
          }
        }
        case "nexauth" -> {
          if (isNexAuthEnabled) {
            Bukkit.getPluginManager().registerEvents(new NexAuthListener(), this);
          }
        }
        case "other" -> {
          Bukkit.getPluginManager().registerEvents(new OtherAuthListener(), this);
        }
        default -> {
          if (isAuthMeEnabled) {
            Bukkit.getPluginManager().registerEvents(new AuthMeListener(), this);
          } else if (isCatSeedLoginEnabled) {
            Bukkit.getPluginManager().registerEvents(new CatSeedLoginListener(), this);
          } else if (isNexAuthEnabled) {
            Bukkit.getPluginManager().registerEvents(new NexAuthListener(), this);
          }
        }
      }
    }
  }


  private void printSupportPluginLoadStatus() {
    if (getMainConfigManager().getConfigData().loggingSupportPluginStatus()) {
      getLogger().info("-----------------");
      if ("zh".equalsIgnoreCase(systemLanguage)) {
        getLogger().info("检测支持插件是否启用: ");
      } else {
        getLogger().info("Check support plugin enable status: ");
      }
      getLogger().info("floodgate: " + isFloodgateEnabled);
      getLogger().info("CMI: " + isCmiEnabled);
      getLogger().info("EssentialsX: " + isEssxEnabled);
      getLogger().info("HuskHomes: " + isHuskHomesEnabled);
      getLogger().info("AuthMe: " + isAuthMeEnabled);
      getLogger().info("CatSeedLogin: " + isCatSeedLoginEnabled);
      getLogger().info("NexAuth: " + isNexAuthEnabled);
      getLogger().info("-----------------");
    }
  }

  private void loadMetrics() {
    Metrics metrics = new Metrics(this, 17107);
    metrics.addCustomChart(new Metrics.SimplePie("basic_plugin", () -> {
      switch (getBasicPlugin()) {
        case CMI -> {
          return "CMI";
        }
        case EssentialsX -> {
          return "EssentialsX";
        }
        case HuskHomes -> {
          return "HuskHomes";
        }
        default -> {
          return "NONE";
        }
      }
    }));
  }

}

File: src/main/kotlin/cc/dsnb/bedrockplayersupport/BedrockPlayerSupport.kt
package cc.dsnb.bedrockplayersupport

import cc.dsnb.bedrockplayersupport.command.*
import cc.dsnb.bedrockplayersupport.config.LangConfig
import cc.dsnb.bedrockplayersupport.config.MainConfig
import cc.dsnb.bedrockplayersupport.form.MainForm
import cc.dsnb.bedrockplayersupport.form.basic.*
import cc.dsnb.bedrockplayersupport.listener.MainTabCompleteListener
import cc.dsnb.bedrockplayersupport.listener.PlayerRespawnListener
import cc.dsnb.bedrockplayersupport.listener.auth.AuthMeListener
import cc.dsnb.bedrockplayersupport.listener.auth.CatSeedLoginListener
import cc.dsnb.bedrockplayersupport.listener.auth.NexAuthListener
import cc.dsnb.bedrockplayersupport.listener.auth.OtherAuthListener
import cc.dsnb.bedrockplayersupport.listener.basic.*
import cc.dsnb.bedrockplayersupport.manager.ConfigManager
import cc.dsnb.bedrockplayersupport.util.UpdateUtil
import com.github.Anon8281.universalScheduler.UniversalScheduler
import com.github.Anon8281.universalScheduler.scheduling.schedulers.TaskScheduler
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer
import org.bstats.bukkit.Metrics
import org.bstats.charts.SimplePie
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import org.geysermc.floodgate.api.FloodgateApi
import java.io.File

/**
 * @author DongShaoNB
 */
class BedrockPlayerSupport : JavaPlugin() {

    companion object {
        lateinit var instance: BedrockPlayerSupport
        lateinit var basicPlugin: BasicPlugin
        lateinit var authPlugin: AuthPlugin
        lateinit var mainConfigManager: ConfigManager<MainConfig>
        lateinit var langConfigManager: ConfigManager<LangConfig>
        lateinit var scheduler: TaskScheduler
        lateinit var languageInUse: String
        lateinit var mainForm: MainForm
        lateinit var basicForm: BasicForm
        lateinit var floodgateApi: FloodgateApi
        val miniMessage = MiniMessage.miniMessage()
        val legacySection = LegacyComponentSerializer.legacySection()
        var isPapiEnabled = false
        const val PREFIX = "[BedrockPlayerSupport] "
    }

    override fun onEnable() {
        instance = this
        scheduler = UniversalScheduler.getScheduler(this)
        floodgateApi = FloodgateApi.getInstance()
        loadConfig()
        setPluginRunningStatus()
        loadFunction()
        loadCommand()
        loadMetrics()
        checkUpdate()
    }

    override fun onDisable() {
        // Don't need to do anything
    }

    private fun loadConfig() {
        mainConfigManager =
            ConfigManager.create(dataFolder.toPath(), "config.yml", MainConfig::class.java).also {
                it.reloadConfig()
                languageInUse = it.getConfigData().language()
            }
        val langDirectory = File(dataFolder, "/lang/")
        if (langDirectory.exists().not()) {
            langDirectory.mkdir()
        }
        ConfigManager.create(langDirectory.toPath(), "default.yml", LangConfig::class.java)
            .reloadConfig()
        if (File(langDirectory, "${languageInUse}.yml").exists().not()) {
            saveResource("lang/${languageInUse}.yml", false)
        }
        langConfigManager =
            ConfigManager.create(
                langDirectory.toPath(),
                "${languageInUse}.yml",
                LangConfig::class.java
            ).also {
                it.reloadConfig()
            }
    }

    private fun setPluginRunningStatus() {
        val pluginManager = Bukkit.getPluginManager()

        val basicPluginNameToEnum = mapOf(
            "cmi" to BasicPlugin.CMI,
            "essentialsx" to BasicPlugin.ESSENTIALS,
            "huskhomes" to BasicPlugin.HUSKHOMES,
            "advancedteleport" to BasicPlugin.ADVANCEDTELEPORT,
            "sunlight" to BasicPlugin.SUNLIGHT
        )
        val configBasicPlugin = mainConfigManager.getConfigData().basicPlugin().lowercase()
        val selectedPlugin = basicPluginNameToEnum[configBasicPlugin]
        basicPlugin = if ("auto" == configBasicPlugin) {
            listOf("CMI", "Essentials", "HuskHomes", "AdvancedTeleport", "SunLight").firstOrNull { plugin ->
                pluginManager.isPluginEnabled(plugin)
            }?.let {
                BasicPlugin.valueOf(it.uppercase())
            } ?: BasicPlugin.NONE
        } else {
            selectedPlugin ?: BasicPlugin.NONE
        }

        val authPluginNameToEnum = mapOf(
            "authme" to AuthPlugin.AUTHME,
            "catseedlogin" to AuthPlugin.CATSEEDLOGIN,
            "nexauth" to AuthPlugin.NEXAUTH,
            "other" to AuthPlugin.OTHER,
            "none" to AuthPlugin.NONE
        )
        val configAuthPlugin = mainConfigManager.getConfigData().authPlugin().lowercase()
        val selectedAuthPlugin = authPluginNameToEnum[configAuthPlugin]
        authPlugin = if ("auto" == configAuthPlugin) {
            listOf("AuthMe", "CatSeedLogin", "NexAuth").firstOrNull { plugin ->
                pluginManager.isPluginEnabled(plugin)
            }?.let {
                AuthPlugin.valueOf(it.uppercase())
            } ?: AuthPlugin.OTHER
        } else {
            selectedAuthPlugin ?: AuthPlugin.NONE
        }

        isPapiEnabled = pluginManager.isPluginEnabled("PlaceholderAPI")
    }

    private fun loadFunction() {
        mainForm = MainForm()
        val pluginManager = Bukkit.getPluginManager()
        val enableRTF = mainConfigManager.getConfigData().enableReceiveTeleportForm()
        when (basicPlugin) {
            BasicPlugin.CMI -> {
                basicForm = CMIForm()
                if (enableRTF) pluginManager.registerEvents(CMIListener(), this)
            }

            BasicPlugin.ESSENTIALS -> {
                basicForm = EssXForm()
                if (enableRTF) pluginManager.registerEvents(EssXListener(), this)
            }

            BasicPlugin.HUSKHOMES -> {
                basicForm = HuskHomesForm()
                if (enableRTF) pluginManager.registerEvents(HuskHomesListener(), this)
            }

            BasicPlugin.ADVANCEDTELEPORT -> {
                basicForm = ATForm()
                if (enableRTF) pluginManager.registerEvents(ATListener(), this)
            }

            BasicPlugin.SUNLIGHT -> {
                basicForm = SunLightForm()
                if (enableRTF) pluginManager.registerEvents(SunLightListener(), this)
            }

            BasicPlugin.NONE -> {
                // Don't need to do anything
            }
        }
        when (authPlugin) {
            AuthPlugin.AUTHME -> pluginManager.registerEvents(AuthMeListener(), this)
            AuthPlugin.CATSEEDLOGIN -> pluginManager.registerEvents(CatSeedLoginListener(), this)
            AuthPlugin.NEXAUTH -> pluginManager.registerEvents(NexAuthListener(), this)
            AuthPlugin.OTHER -> pluginManager.registerEvents(OtherAuthListener(), this)
            AuthPlugin.NONE -> {
                // Don't need to do anything
            }
        }
        if (mainConfigManager.getConfigData().enableBackDeathLocForm()) {
            pluginManager.registerEvents(PlayerRespawnListener(), this)
        }
    }

    private fun loadCommand() {
        val config = mainConfigManager.getConfigData()
        getCommand("bedrockplayersupport")?.also {
            it.setExecutor(MainCommand())
            it.tabCompleter = MainTabCompleteListener()
        }
        if (config.enableTeleportForm()) {
            getCommand("tpgui")?.setExecutor(TpFormCommand())
        }
        if (config.enableMsgForm()) {
            getCommand("msggui")?.setExecutor(MsgFormCommand())
        }
        if (config.enableHomeForm()) {
            getCommand("homegui")?.setExecutor(HomeFormCommand())
        }
        if (config.enablePublicHomeForm()) {
            getCommand("phomegui")?.setExecutor(PublicHomeFormCommand())
        }
        if (config.enableWarpForm()) {
            getCommand("warpgui")?.setExecutor(WarpFormCommand())
        }
        if (config.enableKitForm()) {
            getCommand("kitgui")?.setExecutor(KitFormCommand())
        }
    }

    private fun loadMetrics() {
        Metrics(this, 17107).also {
            it.addCustomChart(SimplePie("basic_plugin") {
                basicPlugin.pluginName
            })
            it.addCustomChart(SimplePie("auth_plugin") {
                authPlugin.pluginName
            })
        }
    }

    private fun checkUpdate() {
        if (mainConfigManager.getConfigData().enableCheckUpdate()) {
            scheduler.runTaskAsynchronously {
                val currentVersion = description.version
                UpdateUtil.getLatestVersion { latestVersion ->
                    if (currentVersion == latestVersion) {
                        if ("zh_cn".equals(languageInUse, ignoreCase = true)) {
                            logger.info("插件是最新版本, 继续保持 ~")
                        } else {
                            logger.info("The plugin is the latest version, keep up ~")
                        }
                    } else {
                        if ("zh_cn".equals(languageInUse, ignoreCase = true)) {
                            logger.info("有新版本可以更新!")
                            logger.info("当前版本: $currentVersion | 最新版本: $latestVersion")
                        } else {
                            logger.info("There is a new version that can be updated!")
                            logger.info("Current version: $currentVersion | Latest version: $latestVersion")
                        }
                    }
                }
            }
        }
    }

}